hw2/hw2.py

```diff
@@ -0,0 +1,420 @@
+
+    Input:
+    - data: any dataset where the last column holds the labels.
+
+    Returns:
+    - entropy: The entropy value.
+    """
+    entropy = 0.0
+    labels = data[:, -1]
+    _, counts = np.unique(labels, return_counts=True)
+    probabilities = counts / counts.sum()
+    entropy = -np.sum(probabilities * np.log2(probabilities))
+    return entropy
+
+
+class DecisionNode:
+
+    def __init__(
+        self,
+        data,
+        impurity_func,
+        feature=-1,
+        depth=0,
+        chi=1,
+        max_depth=1000,
+        gain_ratio=False,
+    ):
+
+        self.data = data  # the relevant data for the node
+        self.feature = feature  # column index of criteria being tested
+        self.pred = self.calc_node_pred()  # the prediction of the node
+        self.depth = depth  # the current depth of the node
+        self.children = []  # array that holds this nodes children
+        self.children_values = []
+        self.terminal = False  # determines if the node is a leaf
+        self.chi = chi
+        self.max_depth = max_depth  # the maximum allowed depth of the tree
+        self.impurity_func = impurity_func
+        self.gain_ratio = gain_ratio
+        self.feature_importance = 0
+
+    def calc_node_pred(self):
+        """
+        Calculate the node prediction.
+
+        Returns:
+        - pred: the prediction of the node
+        """
+        labels = self.data[:, -1]
+        if labels.size == 0:
+            return None  # Handle case with no data in the node
+        unique, counts = np.unique(labels, return_counts=True)
+        return unique[np.argmax(counts)]
+
+    def add_child(self, node, val):
+        """
+        Adds a child node to self.children and updates self.children_values
+
+        This function has no return value
+        """
+        self.children.append(node)
+        self.children_values.append(val)
+
+    def calc_feature_importance(self, n_total_sample):
+        """
+        Calculate the selected feature importance.
+
+        Input:
+        - n_total_sample: the number of samples in the dataset.
+
+        This function has no return value - it stores the feature importance in
+        self.feature_importance
+        """
+        if (
+            self.terminal or self.feature == -1
+        ):  # Terminal or no feature used for splitting
+            self.feature_importance = 0
+            return
+
+        # Impurity of the current node
+        n_samples = self.data.shape[0]  # Number of samples at the current node
+        node_impurity = self.impurity_func(self.data)
+
+        # Calculate weighted impurity for the current node
+        weighted_impurity_current = (n_samples / n_total_sample) * node_impurity
+
+        # Calculate the sum of weighted impurities for each child node
+        weighted_impurities_children = np.sum(
+            [
+                (child.data.shape[0] / n_total_sample) * child.impurity_func(child.data)
```

Comment:
Use - `self.impurity_func` instead of child's.
This is explicitly ensure using same impurity func in calculation, eliminating the question of theoretical scenarios where different func (entropy/gini) is used between parent<->child.